jupyterhub/scopes.py

```diff
@@ -1,18 +1,36 @@
 import functools
 import inspect
+import re
 from enum import Enum
 
 from tornado import web
 from tornado.log import app_log
 
-from . import orm
+from . import roles
 
 
 class Scope(Enum):
     ALL = True
 
 
-def get_user_scopes(name):
+# Used to identify scope filters
+kind_regex = re.compile(r':?(user|service|group)s:?')
+
+
+def build_scope_schema(handler):
+    """Parse raw scope collection into a dict with filters that can be used to resolve API access"""
+    app_log.debug("Parsing scopes")
+    if handler.current_user is not None:
+        handler.raw_scopes = roles.get_subscopes(*handler.current_user.roles)
+    oauth_token = handler.get_current_user_oauth_token()
+    if oauth_token:
+        handler.raw_scopes |= get_user_scopes(oauth_token.name)
```

Comment:
This is performing *unions*, which means that the token can *increase* a user's permissions, which I think we don't want.

I believe the design we have is that a user's tokens can only *limit* permissions, never expand them.

Resolving a request's authorized scopes should look roughly like this:

1. get scopes for the request (APIToken if used, otherwise user); (token -> role -> raw scopes)
2. If requester is a token:
    1. compute *intersection* of token scopes and user scopes (user -> roles -> raw scopes)
    2. Discard any scopes the token has and owner does not
    3. Warn if there were any scopes discarded

If tests are passing, I think we need more tests to ensure that tokens limit scopes, and don't expand them. For example:

1. test that a token with a scope the owner does not have will not actually be able to use the expanded scope
2. test that a token with a limited scope will not be able to use a scope that its owner has that t
3. test that services with the same names as users don't get that user's scope(s) (I see a couple pure string comparisons)

Reply:
Ah, I didn't take the restrictions into account yet. I'll change the behaviour and add some unit tests, since the ones I have now don't test this behaviour as far as I know.

Reply:
This is taken care of now!